mesh/mesh.go

```diff
@@ -500,3 +500,81 @@ func (m *Mesh) GetATXs(atxIds []types.AtxId) (map[types.AtxId]*types.ActivationT
 	}
 	return atxs, mIds
 }
+
+// ActiveSetForLayerConsensusView - returns the active set size that matches the view of the contextually valid blocks in the provided layer
+func (m *Mesh) ActiveSetForLayerConsensusView(layer types.LayerID, layersPerEpoch uint16) (map[string]types.AtxId, error) {
+
+	epoch := layer.GetEpoch(layersPerEpoch)
+	firstLayerOfPrevEpoch := types.LayerID(epoch-1) * types.LayerID(layersPerEpoch)
+
+	// build a map of all blocks on the current layer
+	mp := make(map[types.BlockID]struct{})
+	blocks, err := m.LayerBlockIds(layer)
+	for _, bid := range blocks {
+		mp[bid] = struct{}{}
+	}
+
+	countedAtxs := make(map[string]types.AtxId)
+	penalties := make(map[string]struct{})
+
+	traversalFunc := func(blkh *types.BlockHeader) error {
+
+		// only contextually valid blocks
+		if !m.IsContexuallyValid(blkh.Id) {
+			return nil
+		}
+
+		blk, err := m.GetBlock(blkh.Id)
+		if err != nil {
+			return err
+		}
+
+		// count unique ATXs
+		for _, id := range blk.AtxIds {
+			atx, err := m.GetAtx(id)
+			if err != nil {
+				log.Panic("error fetching atx %v from database -- inconsistent state", id.ShortId()) // TODO: handle inconsistent state
+				return fmt.Errorf("error fetching atx %v from database -- inconsistent state", id.ShortId())
+			}
+
+			// make sure the target epoch is our epoch
+			if atx.TargetEpoch(layersPerEpoch) != epoch {
+				m.With().Debug("atx found, but targeting epoch doesn't match publication epoch",
+					log.String("atx_id", atx.ShortId()),
+					log.Uint64("atx_target_epoch", uint64(atx.TargetEpoch(layersPerEpoch))),
+					log.Uint64("actual_epoch", uint64(epoch)))
+				continue
+			}
+
+			// ignore atx from nodes in penalty
+			if _, exist := penalties[atx.NodeId.Key]; exist {
+				m.With().Debug("ignoring atx from node in penalty",
+					log.String("node_id", atx.NodeId.Key), log.String("atx_id", atx.ShortId()))
+				return nil
```

Comment:
why return and not continue? The other atxs will be from different miners..

Reply:
The case should have been tested...

Reply:
Fixed
Changed the UT so that it will catch this case